### PR TITLE
[SPARK-43254][SQL] Assign a name to the error _LEGACY_ERROR_TEMP_2018

### PR DESCRIPTION
**Pull request by dengziming**
### What changes were proposed in this pull request?
Assign the name `CLASS_UNSUPPORTED_BY_MAP_OBJECTS` to the legacy error class `_LEGACY_ERROR_TEMP_2018`.

### Why are the changes needed?
To assign proper name as a part of activity in SPARK-37935

### Does this PR introduce _any_ user-facing change?
Yes, the error message will include the error class name

### How was this patch tested?
Add a unit test to produce the error from user code.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -170,7 +170,7 @@ object ExpressionEncoder {
    * Function that deserializes an [[InternalRow]] into an object of type `T`. This class is not
    * thread-safe.
    */
-  class Deserializer[T](private val expressions: Seq[Expression])
+  class Deserializer[T](val expressions: Seq[Expression])
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:173]: Could you clarify why do you need this.

**Comment by MaxGekk** [comment]
@dengziming Please, format PR's title as other using tags `[SPARK-XXXXX][SQL] Assign ...`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:173]: I think for testing purpose https://github.com/apache/spark/pull/42939/files#diff-b98c99535d2b28cb47774860d500030e732c244c55b1ac05aead5d1cf1e7a602R2585.

Comment (dengziming) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:173]: Yes, change for testing.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 2]
```diff
@@ -2561,6 +2563,40 @@ class DatasetSuite extends QueryTest
 
     checkDataset(ds.filter(f(col("_1"))), Tuple1(ValueClass(2)))
   }
+
+  test("CLASS_UNSUPPORTED_BY_MAP_OBJECTS when creating dataset") {
```
Comment (dengziming) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2567]: Not sure is this the best place for this test case, currently we don't have test case for creating dataset using `spark.createDatSet`, so I add it to here.

**Review by dengziming — COMMENTED** [review]
@MaxGekk Thanks for your review, I have changed the title, PTAL again.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 3]
```diff
@@ -2561,6 +2563,40 @@ class DatasetSuite extends QueryTest
...
+        df.collect()
+      }
+      val expressions = im.resolveAndBind(df.queryExecution.logical.output,
+        spark.sessionState.analyzer)
+        .createDeserializer().expressions
+
+      // Expression decoding error
+      checkError(
+        exception = exception,
+        errorClass = "_LEGACY_ERROR_TEMP_2151",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2587]: Could you open an JIRA for this, and point out the test, please.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 3 above)
Comment (dengziming) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2587]: Yes, I created SPARK-45213 for this.

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -332,6 +332,11 @@
     ],
     "sqlState" : "22003"
   },
+  "CLASS_UNSUPPORTED_BY_MAP_OBJECTS" : {
+    "message" : [
+      "class `<cls>` is not supported by `MapObjects` as resulting collection."
```
[the remark this answers is not in the document]
Comment (dengziming) [reply common/utils/src/main/resources/error/error-classes.json]: Nice improvement, done.

**Review by dengziming — COMMENTED** [review]
@MaxGekk Thanks for you suggestions, I have resolved them.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
@dengziming Could you rebase on the recent master, please.

**Comment by dengziming** [comment]
The CI passed, ping @MaxGekk

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @dengziming.

**Pull request closed without merging**